orchestrator/src/main/java/com/yahoo/vespa/orchestrator/status/ZookeeperStatusService.java

```diff
@@ -247,6 +285,55 @@ public HostStatus getHostStatus(ApplicationInstanceReference applicationInstance
         }
     }
 
+    /** Do not call this directly: should be called behind a cache. */
+    private HostInfos getHostInfosFromZk(ApplicationInstanceReference application) {
+        Map<HostName, HostInfo> hostInfos;
+        String hostsRootPath = hostsPath(application);
+        if (uncheck(() -> curator.framework().checkExists().forPath(hostsRootPath)) == null) {
+            hostInfos = new HashMap<>();
+        } else {
+            List<String> hostnames = uncheck(() -> curator.framework().getChildren().forPath(hostsRootPath));
+            hostInfos = new HashMap<>(hostnames.stream().collect(Collectors.toMap(
+                    hostname -> new HostName(hostname),
+                    hostname -> {
+                        byte[] bytes = uncheck(() -> curator.framework().getData().forPath(hostsRootPath + "/" + hostname));
+                        return WireHostInfo.deserialize(bytes);
+                    })));
+        }
+
+        // For backwards compatibility we'll add HostInfos from the old hosts-allowed-down ZK path,
+        // using the creation time as the since path. The new hosts ZK path should contain a subset of
+        // the hostnames under hosts-allowed-down ZK path. Eventually these sets should be identical.
+        // Once that's true we can stop writing to hosts-allowed-down, remove this code, and all
+        // data in hosts-allowed-down can be removed.
+        Set<HostName> legacyHostsDown = hostsDownFor(application);
+        Map<HostName, HostInfo> legacyHostInfos = legacyHostsDown.stream().collect(Collectors.toMap(
```

Comment:
Filter for what's not already in `hostInfos`?

Reply:
Done